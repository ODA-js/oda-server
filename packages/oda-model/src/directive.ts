import {
  ModelBase,
  IModelBase,
  ModelBaseInternal,
  ModelBaseInput,
  ModelBaseOutput,
  ModelBaseMetaInfo,
} from './modelbase';
import {
  DirectiveLocation,
  AsHash,
  MetaModelType,
  Nullable,
  assignValue,
  NamedArray,
  MapToArray,
} from './types';
import { inputArgs } from './utils/converters';
import { Internal } from './element';
import { merge, mergeWith } from 'lodash';
import {
  IObjectTypeField,
  ObjectTypeFieldInput,
  ObjectTypeFieldOutput,
} from './objecttypefield';
import { IObjectType, ObjectTypeInput, ObjectTypeOutput } from './objecttype';

export interface IDirective
  extends IModelBase<DirectiveMetaInfo, DirectiveInput, DirectiveOutput> {
  /**
   * set of arguments
   */
  readonly args: Map<string, IObjectTypeField | IObjectType>;
  /**
   * where it can met
   */
  readonly on: Set<DirectiveLocation>;
}

export interface DirectiveMetaInfo extends ModelBaseMetaInfo {}

export interface DirectiveInternal extends ModelBaseInternal {
  args: Map<string, IObjectTypeField | IObjectType>;
  on: Set<DirectiveLocation>;
}

export interface DirectiveInput extends ModelBaseInput<DirectiveMetaInfo> {
  args?:
    | AsHash<ObjectTypeFieldInput | ObjectTypeInput>
    | NamedArray<ObjectTypeFieldInput | ObjectTypeInput>;
  on: DirectiveLocation[];
}

export interface DirectiveOutput extends ModelBaseOutput<DirectiveMetaInfo> {
  args: NamedArray<ObjectTypeOutput | ObjectTypeFieldOutput>;
  on: DirectiveLocation[];
}

export const directiveDefaultMetaInfo = {};
export const directiveDefaultInput = { metadata: directiveDefaultMetaInfo };

export class Directive
  extends ModelBase<
    DirectiveMetaInfo,
    DirectiveInput,
    DirectiveInternal,
    DirectiveOutput
  >
  implements IDirective {
  public get modelType(): MetaModelType {
    return 'directive';
  }
  get args(): Map<string, IObjectTypeField | IObjectType> {
    return this[Internal].args;
  }

  get on(): Set<DirectiveLocation> {
    return this[Internal].on;
  }

  constructor(init: DirectiveInput) {
    super(merge({}, directiveDefaultInput, init));
  }

  public updateWith(input: Nullable<DirectiveInput>) {
    super.updateWith(input);

    assignValue<
      DirectiveInternal,
      DirectiveInput,
      NonNullable<DirectiveInput['args']>
    >({
      src: this[Internal],
      input,
      field: 'args',
      effect: (src, value) => (src.args = inputArgs(value)),
      setDefault: src => (src.args = new Map()),
    });

    assignValue<DirectiveInternal, DirectiveInput, DirectiveInput['on']>({
      src: this[Internal],
      input,
      field: 'on',
      effect: (src, value) => {
        src.on = new Set<DirectiveLocation>(value);
      },
      setDefault: src => (src.on = new Set<DirectiveLocation>()),
    });
  }

  // it get fixed object
  public toObject(): DirectiveOutput {
    return merge({}, super.toObject(), {
      args: MapToArray(this[Internal].args, (_name, value) => value.toObject()),
      on: [...this[Internal].on],
    } as Partial<DirectiveOutput>);
  }

  public mergeWith(payload: Nullable<DirectiveInput>) {
    const update = mergeWith(
      this.toObject(),
      payload,
      (
        o: any,
        s: any,
        key: string,
        _obj: any,
        _source: any,
        _stack: string[],
      ) => {
        if (key === 'on') {
          return [...o, ...s];
        }
        if (key == 'name') {
          return o;
        }
        return;
      },
    );
    this.updateWith(update);
  }
}
