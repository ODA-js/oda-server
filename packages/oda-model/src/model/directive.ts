import clean from '../lib/json/clean';
import {
  FieldArgs,
  DirectiveInput,
  DirectiveStorage,
  MetaModelType,
} from './interfaces';
import { ModelBase } from './modelbase';

export class Directive extends ModelBase {
  public modelType: MetaModelType = 'field';
  protected $obj: DirectiveStorage;

  get args(): FieldArgs[] {
    return this.$obj.args;
  }

  get on(): string[] {
    return this.$obj.on;
  }

  public updateWith(obj: DirectiveInput) {
    if (obj) {
      super.updateWith(obj);
      this.$obj.args = obj.args;
      this.$obj.on = obj.on;
    }
  }

  // it get fixed object
  public toObject() {
    let props = this.$obj;
    let res = super.toObject();
    return clean({
      ...res,
      args: props.args,
      on: props.on,
    });
  }

  // it get clean object with no default values
  public toJSON() {
    let props = this.$obj;
    let res = super.toJSON();
    return clean({
      ...res,
      args: props.args,
      on: props.on,
    });
  }
}
