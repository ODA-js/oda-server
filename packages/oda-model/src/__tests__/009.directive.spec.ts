import 'jest';
import { Directive, directiveDefaultMetaInfo } from '../directive';

describe('Directive', () => {
  it('default', () => {
    const res = new Directive({
      name: 'readOnly',
      on: ['ENUM'],
      args: [
        {
          name: 'message',
          required: false,
          defaultValue: 'THIS PROP IS READ ONLY',
        },
      ],
    });
    expect(res.metadata).toMatchObject(directiveDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('name by default not touched', () => {
    const res = new Directive({ name: 'AbT', on: ['ENUM'] });
    expect(res.name).toBe('AbT');
  });
});
