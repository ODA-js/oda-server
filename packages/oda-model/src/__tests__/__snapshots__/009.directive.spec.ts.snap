// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Directive default: toObject 1`] = `
Object {
  "args": Array [
    Object {
      "description": undefined,
      "metadata": Object {
        "defaultValue": "THIS PROP IS READ ONLY",
        "required": false,
      },
      "multiplicity": "one",
      "name": "message",
      "title": undefined,
      "type": Object {
        "name": "String",
        "type": "scalar",
      },
    },
  ],
  "description": undefined,
  "metadata": Object {},
  "name": "readOnly",
  "on": Array [
    "ENUM",
  ],
  "title": undefined,
}
`;

exports[`Directive merges on: merge 1`] = `
Object {
  "args": Array [],
  "description": undefined,
  "metadata": Object {},
  "name": "Dedupes",
  "on": Array [
    "INTERFACE",
    "FIELD_DEFINITION",
    "FIELD",
  ],
  "title": undefined,
}
`;
